File: SinglyLinkedList.ts
class SinglyLinkedList {
  head: Node | null = null;
  tail: Node | null = null;
  size: number = 0;
  push(val: any) {
    const node = new Node(val);
    if(!this.head) {
      this.head = node;
      this.tail = this.head;
    } else {
      this.tail!.next = node;
      this.tail = this.tail!.next;
    }
    ++this.size;
  }
  reverse() {
    let node = this.head;
    this.head = this.tail;
    this.tail = node;
    let next = null;
    let previous = null;

    for(let i = 0; i !== this.size; ++i) {
      next = node!.next;
      node!.next = previous;
      previous = node;
      node = next;
    }
  }
  print() {
    let arr = [];
    let current = this.head;
    while(current){
      arr.push(current.value);
      current = current.next;
    }
    console.log(arr);
  }
}
class Node {
  value: any;
  next: Node | null = null;
  constructor(val: any) {
    this.value = val;
  }
}

const list = new SinglyLinkedList();

list.push(1)
list.push(2)
list.push(3)
list.push(4)

list.print()
list.reverse()
list.print()

File: SinglyLinkedList.kt
data class Node<T>(
    var value: T,
    var next: Node<T>? = null
)
class SinglyLinkedList<T> {
    private var head: Node<T>? = null
    private var tail: Node<T>? = null
    private var length = 0

    fun push(value: T) {
        val node = Node(value)
        if(head == null) {
            head = node
            tail = head
        } else {
            tail!!.next = node
            tail = tail!!.next
        }
        ++length
    }
    fun reverse() {
        var node = head
        head = tail
        tail = node

        var next: Node<T>?
        var prev: Node<T>? = null

        while(node != null) {
            next = node.next
            node.next = prev
            prev = node
            node = next
        }
    }
    fun print() {
        var current = head
        while (current != null) {
            print(current.value.toString() + " ")
            current = current.next
        }
        println()
    }
}
fun main() {
    val sl = SinglyLinkedList<Number>()
    sl.push(5)
    sl.push(4)
    sl.push(3)
    sl.push(2)
    sl.push(1)
    sl.print()
    sl.reverse()
    sl.print()
}